libc/fuzzing/stdlib/heap_sort_fuzz.cpp

```diff
@@ -0,0 +1,53 @@
+//===-- heap_sort_fuzz.cpp ----------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// Fuzzing test for llvm-libc heap_sort implementation.
+///
+//===----------------------------------------------------------------------===//
+
+#include "src/stdlib/heap_sort.h"
+#include <stdint.h>
+
+static int int_compare(const void *l, const void *r) {
+  int li = *reinterpret_cast<const int *>(l);
+  int ri = *reinterpret_cast<const int *>(r);
+  if (li == ri)
+    return 0;
+  else if (li > ri)
+    return 1;
+  else
+    return -1;
+}
+
+namespace LIBC_NAMESPACE_DECL {
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
```

Comment:
I don't think you should use the namespace here, since it's an `extern "C"` function. qsort_fuzz.cpp doesn't.

Reply:
I tried building without the namespace and using __llvm_libc_19_0_0_git::internal to access the Array object, but when running ninja libc-fuzzer, I was told I had to use LIBC_NAMESPACE_DECL to use __llvm_libc_20_0_0_git instead.

Reply:
You shouldn't use the `__llvm_libc_*` namespace directly, we have the `LIBC_NAMESPACE_DECL` macro for declaring things in the LLVM libc namespace, and the `LIBC_NAMESPACE` macro for using things from the namespace.

This works:

```cpp
  auto arr = LIBC_NAMESPACE::internal::Array(
      reinterpret_cast<uint8_t *>(array), array_size, sizeof(int), int_compare);

  LIBC_NAMESPACE::internal::heap_sort(arr);
```

Reply:
In this case I think it'd be better to just use `LIBC_NAMESPACE::internal::array` instead of putting the whole fuzzer inside  the namespace

Reply:
By the way, according to the C++ standard, accessing the `int` elements of `array` through `reinterpret_cast<uint8_t *>(array)` is undefined behavior.

https://eel.is/c++draft/basic.lval#11

But I guess that in practice, when you compile with Clang, GCC, or MSVC, it always works as you would expect it to, so I don't know if LLVM maintainers care about this.

Reply:
Ah, that's the macro I was looking for, thank you. As for the reinterpret casting, I am matching the qsort fuzzer implementation so it should be fine.

Reply:
What libc/src/stdlib/qsort.cpp does is also undefined behavior. The question is whether LLVM maintainers care, as in practice, with the compilers and targets we support, I think it always does the we expect it to.

Reply:
I asked on the LLVM Discord server and apparently this should be fixed: https://discord.com/channels/636084430946959380/636732535434510338/1267815152678600705.

But it would require changes to existing code that this PR is meant to only a test for, so I don't think it should be done in this PR.